### ISSUE TITLE
FP16 tests fail on Pascal

### ISSUE THREAD

**Issue by borisfom** [soumith/cudnn.torch#225]
I started this issue to track debugging changes from https://github.com/soumith/cudnn.torch/pull/206.
OK I was able to run the test on Tesla card. 
Looks like some sizes (and we use random on the test) may not be supported for backward methods, I will get more info on that. What seem to be straightforward is that backward descriptors should be allocated lazily only if they are actually needed:

Running test on device: #2 : Tesla P100-SXM2-16GB

Testing torch.CudaHalfTensor
Running 37 tests
1/37 VolumetricMaxPooling_batch ........................................ [PASS]
2/37 SpatialMaxPooling_single .......................................... [PASS]
3/37 SpatialCrossEntropyCriterion ...................................... [PASS]
4/37 SpatialMaxPooling_batch ........................................... [PASS]
5/37 SpatialAveragePooling_batch ....................................... [PASS]
6/37 LogSoftMax_single ................................................. [PASS]
7/37 functional_bias2D ................................................. [PASS]
8/37 SpatialBatchNormalization ......................................... [PASS]
9/37 SpatialConvolution_forward_single ................................. [ERROR]
10/37 VolumetricMaxPooling_single ....................................... [PASS]
11/37 SpatialFullConvolution_backward_batch ............................. [ERROR]
12/37 TemporalConvolution_padding_batch ................................. [ERROR]
13/37 SpatialConvolution_backward_batch ................................. [ERROR]
14/37 Tanh_single ....................................................... [PASS]
15/37 ClippedReLU_single ................................................ [PASS]
16/37 ReLU_single ....................................................... [PASS]
17/37 Sigmoid_single .................................................... [PASS]
18/37 VolumetricConvolution_forward_single .............................. [ERROR]
19/37 Tanh_batch ........................................................ [PASS]
20/37 SoftMax_batch ..................................................... [PASS]
21/37 SpatialCrossMapLRN_batch .......................................... [PASS]
22/37 VolumetricBatchNormalization ...................................... [PASS]
23/37 SpatialConvolution_backward_single ................................ [ERROR]
24/37 ReLU_batch ........................................................ [PASS]
25/37 TemporalConvolution_single ........................................ [ERROR]
26/37 TemporalConvolution_batch ......................................... [ERROR]
27/37 Sigmoid_batch ..................................................... [PASS]
28/37 SpatialLogSoftMax ................................................. [PASS]
29/37 TemporalConvolution_reduceBatchSize ............................... [ERROR]
30/37 LogSoftMax_batch .................................................. [PASS]
31/37 BatchNormalization ................................................ [PASS]
32/37 SpatialFullConvolution_forward_batch .............................. [ERROR]
33/37 functional_maxpooling2d ........................................... [PASS]
34/37 SpatialConvolution_forward_batch .................................. [ERROR]
35/37 SpatialAveragePooling_single ...................................... [PASS]
36/37 VolumetricConvolution_backward_single ............................. [ERROR]
37/37 SoftMax_single .................................................... [PASS]

Completed 173 asserts in 37 tests with 0 failures and 12 errors

SpatialConvolution_forward_single
Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_NOT_SUPPORTED (cudnnGetConvolutionBackwardDataWorkspaceSize)
stack traceback:
[C]: in function 'error'
/usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
/usr/share/lua/5.1/cudnn/SpatialConvolution.lua:313: in function 'createIODescriptors'
/usr/share/lua/5.1/cudnn/SpatialConvolution.lua:364: in function 'forward'
test/test.lua:176: in function 'test'
test/test.lua:189: in function 
[C]: in function 'xpcall'
/usr/share/lua/5.1/torch/Tester.lua:477: in function '_pcall'
/usr/share/lua/5.1/torch/Tester.lua:436: in function '_run'
/usr/share/lua/5.1/torch/Tester.lua:355: in function 'run'
test/test.lua:1500: in main chunk
[C]: in function 'dofile'
/usr/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
[C]: at 0x00406670

SpatialFullConvolution_backward_batch
Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_NOT_SUPPORTED (cudnnGetConvolutionBackwardDataWorkspaceSize)
stack traceback:
[C]: in function 'error'
/usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
/usr/share/lua/5.1/cudnn/SpatialFullConvolution.lua:283: in function 'createIODescriptors'
/usr/share/lua/5.1/cudnn/SpatialFullConvolution.lua:310: in function 'forward'

**Comment by borisfom** [comment]
Those are actually legitimate CUDNN failures, and won't be fixed in upcoming 5.1 release.
@soumith: should we use 'less random' sizes for the test ? Partially working test is better than none... I can't really decipher which combinations fail though, and I can't get definite input from CUDNN.
I have added some instrumentation to tests so that now you can see which combinations fail with cudnn.verbose=true. I have checked an obvious guess - using all even numbers does not fix it.

**Comment by soumith** [comment]
@borisfom cant you just ask the cudnn engineers for the spec? the tests are based on the cudnn manual. If FP16 doesn't support certain sizes on pascal, they should (and probably will) document it somewhere. Once that spec is given to me, i can modify the tests accordingly.

**Comment by borisfom** [comment]
OK I have asked and here's the news:
- For the sizes that fail with FP16, CUDNN has CUDA C fallback for float and double. 
- SpatialConvolution:  change filter size to be <10 in both dimensions i.e. setting it to (say) 64x2x9x5 should work.
- TemporalConvolution uses obscene filter sizes, so it probably won't work with fp16. If filter has to be less than 10x10, than its hard to even artificially cook a test size that would pass, so perhaps TemporalConvolution tests and TemporalConvolution itself should be disabled for fp16. 
- setting strides (si,sj,[sk]) to 1 should help

I have tried implementing some of those in https://github.com/soumith/cudnn.torch/pull/237 (e.g see random2 function), but I still do have failures from time to time due to randomization. 
@soumith , @szagoruyko, @ngimel  : you could probably do it better. Also, if we want to have the right performance feel and not testing fallbacks, we should implements same test sizes for all types ?

**Comment by soumith** [comment]
Can you please find out if CuDNN will not have FP16 support for larger than 9x9 kernels? There's several research projects that use this, not just via TemporalConvolution, but SpatialConvolution as well (Jonathan Tompson's papers use 128x128 kernels). Even providing a simple not-great-perf kernel would go a long way.

**Comment by ngimel** [comment]
Are you more interested in pseudo-fp16 (fp16 i/o, fp32 math) or true fp16? Current support and future plans are different for those two cases.

**Comment by soumith** [comment]
> Are you more interested in pseudo-fp16 (fp16 i/o, fp32 math) or true fp16? 

At this point we're interested in Maxwell, so pseudo-fp16. But once pascal deployments happen, that might change.

**Comment by soumith** [comment]
but more importantly, cudnn so far gave the guarantee of handling all cases.
If we will have to deal with fallbacks at the framework level, then I'll have to rewrite the bindings slightly to have proper fallbacks to nn where appropriate.

**Comment by borisfom** [comment]
@natalia: I believe true vs pseudo CUDNN modes are going to be pretty transparent from Torch side, right? In all cases cutorch would feed CUDNN the same data and receive the same data. The only difference I may foresee is performance of CUDNN call (up to 2x faster when true fp16 is used, but even with fp32 it would beat the fallback) and the fact that CUDNN may chose to operate with 32-bit workspace in case pseudo - cudnn.torch would see it as different memory sizes returned from FindEx call.

**Comment by ngimel** [comment]
@borisfom, true fp16 and pseudo fp16 have different parameters of convolution descriptor. Though we can abstract it on the torch side, I'm not sure that we should - after all, accuracy of fp16 computation can be different from accuracy achieved with fp32. So even on cards that have true fp16, it makes sense to leave an option of fp32 math. Similarly to how additional algos are on one hand transparent behind cudnn.fastest and cudnn.benchmark options, but also can be set explicitly. 
@soumith, cudnn team says that for pseudo-fp16 all sizes should be supported already, for true fp16 full support is coming in cudnn v6.

**Comment by borisfom** [comment]
Thanks @ngimel! No we should not abstract it and keep it transparent, but I believe an option to allow fallback should be there. 
@soumith: how do you think we should go about it ? I am thinking, instead of cudnn flag, an assignable callback makes more sense: we can pass the whole current convolution to the handler. Default hander would print a warning and return status that would make the algo chooser to fall back to 32 bits in case of failed 16-bit convolution, and to error out otherwise.

**Comment by borisfom** [comment]
@soumith: I am adding generic fallback to cudnn.find, like : 
find.float16to32fallback  = function(layer)
--- reads currrent conv descriptor and resets with float type if it's float16 ...
float16to32fallback can be reassigned.

**Comment by borisfom** [comment]
@soumith: please take a look at https://github.com/soumith/cudnn.torch/pull/250/commits/5afec3b4ee65bf0683e0c6831cf758a9f8b15e2a
It works in most cases, but yet again not always and that bothers me ...

**Comment by borisfom** [comment]
Update: 
1. Fallback to pseudo-fp16 is implemented in https://github.com/soumith/cudnn.torch/pull/250 - only triggered for some sizes in 2d case, and 3d convolutions are always pseudo.
For CUDNN5, those limitations will persists so this is the best we can do.
2. Also, @ngimel and myself agreed that for current state of art , most reasonable default setting is to force pseudo-fp16 on Pascal, as per https://github.com/torch/cutorch/pull/521 and https://github.com/soumith/cudnn.torch/pull/267
